x/pairing/keeper/scores/score.go

```diff
@@ -0,0 +1,216 @@
+// 2. Generate pairing slots with requirements (one slot per provider).
+// 3. Compute the pairing score of each provider with respect to each slot.
+// 4. Pick a provider for each slot with a pseudo-random weighted choice.
+//
+// Pairing requirements describe the policy-imposed requirements for paired providers. Examples include
+// geolocation constraints, ability to service archive requests, and expectations regarding QoS ranking
+// of selected providers. Pairing requirements must satisfy the ScoreReq interface, whose methods can
+// identify the ScoreReq (by name), and compute a score for a provider with respect to that requirement.
+//
+// A pairing slot represents a single provider slot in the pairing list (The number of slots for pairing
+// is defined by the policy). Each pairing slot holds a set of pairing requirements (a pairing slot may
+// repeat). For example, a policy may state that the pairing list has 6 slots, and providers should be
+// located in Asia and Europe. This can be satisfied with a pairing list that has 3 (identical) slots
+// that require providers in Asia and 3 (identical) slots that require providers in Europe.
+//
+// A pairing score describes the suitability of a provider for a pairing slot (under a given strategy).
+// The score depends on the slot's requirements: for example, given a slot which requires geolocation in
+// Asia, a provider in Asia will generally get higher score than one in Europe. The score is calculated for
+// each <provider, slot> combination.
+//
+// A pairing score strategy defines the weight of each score requirement in the final score calculation
+// for a <provider, slot> combination. For example, given a slot with several requirements, then the
+// overall pairing score would be calculated as score1^w1 + score2^w2 + ... (where score1 is the score
+// of the provider with respect to the first requirement, score2 with respect to the second requirement
+// and so on).
+//
+//
+// To add a new requirement, one should create a new object that satisfies the ScoreReq interface and update the
+// CalcSlots() function so the new requirement will be assigned to the pairing slots (according to some logic).
+// Lastly, append the new requirement object to the allReqTypes var in the init() function in this file. Use
+// StakeReq or GeoReq as examples.
+
+package scores
+
+import (
+	"fmt"
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	commontypes "github.com/lavanet/lava/common/types"
+	epochstoragetypes "github.com/lavanet/lava/x/epochstorage/types"
+	planstypes "github.com/lavanet/lava/x/plans/types"
+	tendermintcrypto "github.com/tendermint/tendermint/crypto"
+)
+
+var (
+	uniformStrategy ScoreStrategy
+	allReqNames     []string
+)
+
+// TODO: currently we'll use weight=1 for all reqs. In the future, we'll get it from policy
+func init() {
+	// gather all req names to a list
+	allReqNames = append(allReqNames, stakeReqName)
```

Comment:
but I want this variable to be available outside of the init() function. So I declare it above and append. In your way, its scope will be init()

Reply:
why would the scope be within init?
the variable is declared outside of it, just looks weird appending an empty list instead of initializing it with the values.
this is not a must, just my opinion

Reply:
I thought I saw allReqName := ... (with ":"). see https://github.com/lavanet/lava/commit/6ba55d5846fd3f875247146f357d52c5249e7d4a